libraries/SITL/SIM_GPS.cpp

```diff
@@ -1018,6 +1043,234 @@ uint32_t GPS::CalculateBlockCRC32(uint32_t length, uint8_t *buffer, uint32_t crc
+    struct PACKED gsof_pos_sigma {
+        const uint8_t OUTPUT_RECORD_TYPE = GSOF_POS_SIGMA_TYPE; 
+        const uint8_t RECORD_LEN = GSOF_POS_SIGMA_LEN;
+        uint32_t pos_rms = htobe32(0);
+        uint32_t sigma_e = htobe32(0);
+        uint32_t sigma_n = htobe32(0);
+        uint32_t cov_en = htobe32(0);
+        uint32_t sigma_up = htobe32(0);
+        uint32_t semi_major_axis = htobe32(0);
+        uint32_t semi_minor_axis = htobe32(0);
+        uint32_t orientation = htobe32(0);
+        uint32_t unit_variance = htobe32(0);
+        uint16_t n_epocs = htobe32(1); // Always 1 for kinematic.
+    } pos_sigma {};
+    [[maybe_unused]] const auto sps = sizeof(gsof_pos_sigma);
+    static_assert(sizeof(gsof_pos_sigma) - (sizeof(gsof_pos_sigma::OUTPUT_RECORD_TYPE) + sizeof(gsof_pos_sigma::RECORD_LEN)) == GSOF_POS_SIGMA_LEN);
+    
+    // TODO add GSOF49
+
+    send_gsof(GSOF_POS_TIME_TYPE, (uint8_t*)&pos_time, sizeof(pos_time));
+    send_gsof(GSOF_POS_TYPE, (uint8_t*)&pos, sizeof(pos));
+    send_gsof(GSOF_VEL_TYPE, (uint8_t*)&vel, sizeof(vel));
+    send_gsof(GSOF_DOP_TYPE, (uint8_t*)&dop, sizeof(dop));
+    send_gsof(GSOF_POS_SIGMA_TYPE, (uint8_t*)&pos_sigma, sizeof(pos_sigma));
+}
+
+
+void GPS::send_gsof(uint8_t output_record_type, uint8_t *buf, uint16_t size)
+{
+    // All Trimble "Data Collector" packets, including GSOF, are comprised of three fields:
+    // * A fixed-length packet header (dcol_header)
+    // * A variable-length data frame (buf)
+    // * A fixed-length packet trailer (dcol_trailer)
+    // Reference: // https://receiverhelp.trimble.com/oem-gnss/index.html#API_DataCollectorFormatPacketStructure.html?TocPath=API%2520Documentation%257CData%2520collector%2520format%2520packets%257CData%2520collector%2520format%253A%2520packet%2520structure%257C_____0
+
+    const uint8_t STX = 0x02; 
+    // status bitfield
+    // https://receiverhelp.trimble.com/oem-gnss/index.html#API_ReceiverStatusByte.html?TocPath=API%2520Documentation%257CData%2520collector%2520format%2520packets%257CData%2520collector%2520format%253A%2520packet%2520structure%257C_____1
+    const uint8_t STATUS = 0xa8;
+    const uint8_t PACKET_TYPE = 0x40; // Report Packet 40h (GENOUT)
+
+    // Before writing the GSOF data buffer, the GSOF header needs added between the DCOL header and the payload data frame.
+    // https://receiverhelp.trimble.com/oem-gnss/index.html#GSOFmessages_GSOF.html?TocPath=Output%2520Messages%257CGSOF%2520Messages%257C_____2
+
+    static uint8_t TRANSMISSION_NUMBER = 0; // Functionally, this is a sequence number
+    // Most messages, even GSOF49, only take one page. For SIM, assume it.
+    assert(size < 0xFA); // GPS SIM doesn't yet support paging
+    constexpr uint8_t PAGE_INDEX = 0; 
+    constexpr uint8_t MAX_PAGE_INDEX = 0;
+    [[maybe_unused]] const uint8_t gsof_header[3] = {
+        TRANSMISSION_NUMBER,
+        PAGE_INDEX,
+        MAX_PAGE_INDEX,
+
+    };
+    ++TRANSMISSION_NUMBER;
```

Comment:
Not clear how to handle this with simulated glitches and similar behavior. This isn't even used in the GSOF driver, but I think it should be.

Reply:
Ignoring for initial merge.